File: BehavioralDesignPattern/MementoPattern/java/CareTaker.java
package BehavioralDesignPattern.MementoPattern.java;

import java.util.Stack;

public class CareTaker {
    private Stack<Memento> memstack;

    public CareTaker(){
        this.memstack = new Stack<>();
    }

    public void save(TextEditor textEditor){
        memstack.push(textEditor.save());
    }

    public void undo(TextEditor textEditor){
        if(!memstack.empty()){
            textEditor.restore(memstack.pop());
        }
        else {
            System.out.println("No state to undo");
        }
    }
}

File: BehavioralDesignPattern/MementoPattern/python/CareTaker.py
class CareTaker:
    def __init__(self):
        self.memstack = []
    
    def save(self, texteditor):
        self.memstack.append(texteditor.save())

    def undo(self, texteditor):
        if len(self.memstack) != 0:
            texteditor.restore(self.memstack.pop())
        else:
            print("No state to undo")